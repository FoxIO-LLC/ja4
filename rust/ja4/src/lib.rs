// Copyright (c) 2023, FoxIO, LLC.
// All rights reserved.
// Patent Pending
// JA4 is Open-Source, Licensed under BSD 3-Clause
// JA4+ (JA4S, JA4H, JA4L, JA4X, JA4SSH) are licenced under the FoxIO License 1.1.
// For full license text, see the repo root.

mod conf;
mod error;
mod http;
mod pcap;
mod ssh;
mod stream;
mod time;
mod tls;

use std::{io::Write, path::PathBuf};

use clap::Parser;
use rtshark::RTSharkBuilder;

pub use crate::error::Error;
use crate::{
    conf::Conf,
    pcap::{Packet, PacketNum, Proto},
    stream::Streams,
};

pub type Result<T, E = Error> = std::result::Result<T, E>;

/// Calculate JA4 fingerprints
#[derive(Debug, Parser)]
#[command(version = env!("CARGO_PKG_VERSION"))]
pub struct Cli {
    /// JSON output (default is YAML)
    #[arg(short, long)]
    json: bool,
    /// Include raw (unhashed) fingerprints in the output
    #[arg(short = 'r', long)]
    with_raw: bool,
    /// Preserve the original order of values.
    ///
    /// JA4 (TLS client): disable sorting of ciphers and TLS extensions.
    ///
    /// JA4H (HTTP client): disable sorting of headers and cookies.
    #[arg(short = 'O', long)]
    original_order: bool,
    /// The key log file that enables decryption of TLS traffic.
    ///
    /// This file is generated by the browser when `SSLKEYLOGFILE` environment variable is set.
    /// See <https://wiki.wireshark.org/TLS#using-the-pre-master-secret> for more details.
    ///
    /// Note that you can embed the TLS key log file in a capture file:
    /// `editcap --inject-secrets tls,keys.txt in.pcap out-dsb.pcapng`
    #[arg(long)]
    keylog_file: Option<PathBuf>,
    /// Include packet numbers (`pkt_*` fields) in the output.
    ///
    /// This information is useful for debugging.
    #[arg(short = 'n', long)]
    with_packet_numbers: bool,
    /// The capture file to process
    pcap: PathBuf,
}

impl Cli {
    /// Write JSON with JA4 fingerprints to the I/O stream.
    pub fn run<W: Write>(self, writer: &mut W) -> Result<()> {
        let conf = Conf::load()?;
        let Cli {
            json,
            with_raw,
            original_order,
            keylog_file,
            with_packet_numbers,
            pcap,
        } = self;

        let Some(pcap_path) = pcap.to_str() else {
            return Err(Error::NonUtf8Path(pcap));
        };
        check_tshark_version()?;
        let mut builder = RTSharkBuilder::builder().input_path(pcap_path);

        if let Some(keylog) = &keylog_file {
            let Some(keylog_path) = keylog.to_str() else {
                // SAFETY: we've just established that `keylog_file` is Some
                return Err(Error::NonUtf8Path(keylog_file.unwrap()));
            };
            builder = builder.keylog_file(keylog_path);
        }
        let mut tshark = builder.spawn()?;

        let mut streams = Streams::default();

        let mut packet_num = 0;
        while let Some(packet) = tshark.read().unwrap_or_else(|err| {
            tracing::error!(%err, "failed to parse tshark output");
            None
        }) {
            packet_num += 1;
            let pkt = Packet::new(&packet, packet_num);

            if let Err(error) = streams.update(&pkt, &conf, with_packet_numbers) {
                tracing::debug!(packet_num, %error, "failed to handle packet");
            }
        }

        let flags = FormatFlags {
            with_raw,
            original_order,
        };
        // HACK: The purpose of the `io::stdout` mumbo-jumbo is to handle
        // BrokenPipe error. Rust throws it when the stdout is piped to `head`.
        if json {
            for rec in streams.into_out(flags) {
                serde_json::to_writer(&mut *writer, &rec)?;
                writeln!(writer)?;
            }
        } else {
            let s = serde_yaml::to_string(&streams.into_out(flags).collect::<Vec<_>>())?;
            writer.write_all(s.as_bytes())?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, Copy, Default)]
pub(crate) struct FormatFlags {
    /// Whether to add raw (unhashed) fingerprints to the output.
    ///
    /// Corresponds to [`Cli::with_raw`].
    pub(crate) with_raw: bool,
    /// Whether to preserve the original order of values.
    ///
    /// Corresponds to [`Cli::original_order`].
    pub(crate) original_order: bool,
}

/// Which side of the connection sent the packet?
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub(crate) enum Sender {
    Client,
    Server,
}

/// Returns first 12 characters of the SHA-256 hash of the given string.
///
/// Returns `"000000000000"` (12 zeros) if the input string is empty.
fn hash12(s: impl AsRef<str>) -> String {
    use sha2::{Digest as _, Sha256};

    let s = s.as_ref();
    if s.is_empty() {
        "000000000000".to_owned()
    } else {
        let sha256 = hex::encode(Sha256::digest(s));
        sha256[..12].into()
    }
}

#[test]
fn test_hash12() {
    assert_eq!(hash12("551d0f,551d25,551d11"), "aae71e8db6d7");
    assert_eq!(hash12(""), "000000000000");
}

fn check_tshark_version() -> Result<()> {
    use owo_colors::OwoColorize as _;

    let out = duct::cmd!("tshark", "--version")
        .read()
        .map_err(|e| Error::TsharkNotFound { source: e })?;
    tracing::debug!(%out, "tshark --version");

    let ver = parse_tshark_version(&out).ok_or(Error::ParseTsharkVersion)?;
    let available = semver::Version::parse(ver)?;

    let required = semver::VersionReq::parse(">=4.0.6").expect("BUG");
    if !required.matches(&available) {
        tracing::warn!(%available, %required, "tshark version is outdated");
        let warning = format!(
            "⚠️  You are running an older version of tshark ({available}).\n\
            JA4 is designed to work with tshark version 4.0.6 and above.\n\
            Some functionality may not work properly with older versions."
        );
        eprintln!("{}", warning.bold().red());
    }
    Ok(())
}

/// Parses the version number from the output of `tshark --version`.
fn parse_tshark_version(tshark_version_output: &str) -> Option<&str> {
    // The first line of `tshark --version` output is formatted like this:
    // "TShark (Wireshark) 4.0.8 (v4.0.8-0-g81696bb74857).\n"
    let start = tshark_version_output.find(") ").map(|i| i + 2)?;
    let version_start = &tshark_version_output[start..];
    let end = version_start.find(char::is_whitespace)?;
    let ver = &version_start[..end];
    Some(ver.strip_suffix('.').unwrap_or(ver))
}

#[test]
fn test_parse_tshark_version() {
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 4.0.8 (v4.0.8-0-g81696bb74857)."),
        Some("4.0.8")
    );
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 3.6.2 (Git v3.6.2 packaged as 3.6.2-2)"),
        Some("3.6.2")
    );
    assert_eq!(
        parse_tshark_version("TShark (Wireshark) 4.4.0.\n\nCopyright 1998-2024"),
        Some("4.4.0")
    );
    // Abrupt end of the string.
    assert!(parse_tshark_version("TShark (Wireshark) 4.4.0.").is_none());
    assert!(parse_tshark_version("What the TShark?!").is_none());
}

// XXX-FIXME(vvv): `test_insta` fails on Windows; see https://github.com/FoxIO-LLC/ja4/issues/10
#[cfg(not(windows))]
#[test]
fn test_insta() {
    insta::glob!(
        concat!(env!("CARGO_MANIFEST_DIR"), "/../.."),
        "pcap/*.pcap*",
        |path| {
            let cli = Cli {
                json: false,
                with_raw: false,
                original_order: false,
                keylog_file: None,
                with_packet_numbers: false,
                pcap: path.to_path_buf(),
            };

            let mut output = Vec::<u8>::new();
            cli.run(&mut output).unwrap();
            let output = String::from_utf8(output).unwrap();

            insta::assert_snapshot!(output);
        }
    );
}
