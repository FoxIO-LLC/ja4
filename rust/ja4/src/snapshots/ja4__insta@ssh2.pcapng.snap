---
source: ja4/src/lib.rs
expression: output
---
- stream: 5
  transport: tcp
  src: 172.16.225.48
  dst: 146.112.255.155
  src_port: 57368
  dst_port: 443
  tls_server_name: updates.opendns.com
  ja4: t12d190800_d83cc789557e_7af1ed941c26
  ja4s: t120400_c02f_4993ccf7354b
  tls_certs:
  - x509:
    - ja4x: a373a9f83c6b_2bab15409345_7bf9a7bf7029
      issuerCountryName: US
      issuerOrganizationName: DigiCert Inc
      issuerCommonName: DigiCert TLS RSA SHA256 2020 CA1
      subjectCountryName: US
      subjectStateOrProvinceName: California
      subjectLocalityName: San Francisco
      subjectOrganizationName: Cisco OpenDNS LLC
      subjectCommonName: api.opendns.com
    - ja4x: 7d5dbb3783b4_a373a9f83c6b_a83ffcd6e6c2
      issuerCountryName: US
      issuerOrganizationName: DigiCert Inc
      issuerOrganizationalUnit: www.digicert.com
      issuerCommonName: DigiCert Global Root CA
      subjectCountryName: US
      subjectOrganizationName: DigiCert Inc
      subjectCommonName: DigiCert TLS RSA SHA256 2020 CA1
  ja4l_c: 44_128
  ja4l_s: 12517_53
- stream: 8
  transport: tcp
  src: 172.16.225.48
  dst: 34.248.242.11
  src_port: 57371
  dst_port: 443
  tls_server_name: mcs2-cloudstation-eu-west-1.prod.hydra.sophos.com
  ja4: t12d1909h2_d83cc789557e_7af1ed941c26
  ja4s: t120500_c02f_6471ab80eb72
  tls_certs:
  - x509:
    - ja4x: f7a0b866a27b_30b9c68c9fc8_8f2dd91f85ae
      issuerCommonName: Sophos SHA256 MCS Root CA3
      issuerStateOrProvinceName: Oxfordshire
      issuerCountryName: UK
      issuerOrganizationName: Sophos Ltd
      subjectCommonName: mcs2-cloudstation-eu-west-1.prod.hydra.sophos.com
      subjectStateOrProvinceName: Oxfordshire
      subjectCountryName: GB
      subjectOrganizationName: Sophos Ltd.
      subjectOrganizationalUnit: SaaS
    - ja4x: f7a0b866a27b_f7a0b866a27b_b189698ac141
      issuerCommonName: Sophos SHA256 MCS Root CA3
      issuerStateOrProvinceName: Oxfordshire
      issuerCountryName: UK
      issuerOrganizationName: Sophos Ltd
      subjectCommonName: Sophos SHA256 MCS Root CA3
      subjectStateOrProvinceName: Oxfordshire
      subjectCountryName: UK
      subjectOrganizationName: Sophos Ltd
  ja4l_c: 56_128
  ja4l_s: 55492_235
- stream: 11
  transport: tcp
  src: 172.16.225.48
  dst: 52.178.17.3
  src_port: 57374
  dst_port: 443
  tls_server_name: self.events.data.microsoft.com
  ja4: t12d190800_d83cc789557e_7af1ed941c26
  ja4s: t120300_c030_09f674154ab3
  tls_certs:
  - x509:
    - ja4x: a373a9f83c6b_2bab15409345_0f2217ba412e
      issuerCountryName: US
      issuerOrganizationName: Microsoft Corporation
      issuerCommonName: Microsoft Azure TLS Issuing CA 05
      subjectCountryName: US
      subjectStateOrProvinceName: WA
      subjectLocalityName: Redmond
      subjectOrganizationName: Microsoft Corporation
      subjectCommonName: '*.events.data.microsoft.com'
    - ja4x: 7d5dbb3783b4_a373a9f83c6b_c34b04c10969
      issuerCountryName: US
      issuerOrganizationName: DigiCert Inc
      issuerOrganizationalUnit: www.digicert.com
      issuerCommonName: DigiCert Global Root G2
      subjectCountryName: US
      subjectOrganizationName: Microsoft Corporation
      subjectCommonName: Microsoft Azure TLS Issuing CA 05
  ja4l_c: 46_128
  ja4l_s: 49308_110
- stream: 12
  transport: tcp
  src: 172.16.225.48
  dst: 204.79.197.220
  src_port: 57375
  dst_port: 443
  tls_server_name: www.bing.com
  ja4: t13d1516h2_8daaf6152771_e5627efa2ab1
  ja4s: t1205h2_c030_015e35fdd027
  tls_certs:
  - x509:
    - ja4x: a373a9f83c6b_7022c563de38_0ce9ea683d50
      issuerCountryName: US
      issuerOrganizationName: Microsoft Corporation
      issuerCommonName: Microsoft RSA TLS CA 02
      subjectCommonName: www.bing.com
    - ja4x: 7d5dbb3783b4_a373a9f83c6b_44440d41940c
      issuerCountryName: IE
      issuerOrganizationName: Baltimore
      issuerOrganizationalUnit: CyberTrust
      issuerCommonName: Baltimore CyberTrust Root
      subjectCountryName: US
      subjectOrganizationName: Microsoft Corporation
      subjectCommonName: Microsoft RSA TLS CA 02
  ja4l_c: 55_128
  ja4l_s: 3217_119
- stream: 13
  transport: tcp
  src: 172.16.225.48
  dst: 52.86.25.233
  src_port: 57376
  dst_port: 443
  tls_server_name: 4.sophosxl.net
  ja4: t12d1909h2_d83cc789557e_7af1ed941c26
  ja4s: t120200_c02f_344b4dce5a52
  tls_certs:
  - x509:
    - ja4x: a373a9f83c6b_7022c563de38_7bf9a7bf7029
      issuerCountryName: US
      issuerOrganizationName: Amazon
      issuerCommonName: Amazon RSA 2048 M02
      subjectCommonName: 4.sophosxl.net
    - ja4x: a373a9f83c6b_a373a9f83c6b_3684c5172069
      issuerCountryName: US
      issuerOrganizationName: Amazon
      issuerCommonName: Amazon Root CA 1
      subjectCountryName: US
      subjectOrganizationName: Amazon
      subjectCommonName: Amazon RSA 2048 M02
    - ja4x: 2bab15409345_a373a9f83c6b_44ce05048d28
      issuerCountryName: US
      issuerStateOrProvinceName: Arizona
      issuerLocalityName: Scottsdale
      issuerOrganizationName: Starfield Technologies, Inc.
      issuerCommonName: Starfield Services Root Certificate Authority - G2
      subjectCountryName: US
      subjectOrganizationName: Amazon
      subjectCommonName: Amazon Root CA 1
    - ja4x: e7bc7ebc3d9e_2bab15409345_44ce05048d28
      issuerCountryName: US
      issuerOrganizationName: Starfield Technologies, Inc.
      issuerOrganizationalUnit: Starfield Class 2 Certification Authority
      subjectCountryName: US
      subjectStateOrProvinceName: Arizona
      subjectLocalityName: Scottsdale
      subjectOrganizationName: Starfield Technologies, Inc.
      subjectCommonName: Starfield Services Root Certificate Authority - G2
  ja4l_c: 58_128
  ja4l_s: 18693_241
- stream: 14
  transport: tcp
  src: 172.16.225.48
  dst: 54.160.114.75
  src_port: 57377
  dst_port: 22
  ja4l_c: 77_128
  ja4l_s: 12897_50
  ja4ssh:
  - c36s36_c76s124_c74s5
  - c36s52_c42s76_c51s2
  ssh_extras:
    hassh: 06046964c022c6407d15a27b12a6a4fb
    hassh_server: 699519fdcc30cbcd093d5cd01e4b1d56
    ssh_protocol_client: SSH-2.0-OpenSSH_7.6p1 Ubuntu-4ubuntu0.5
    ssh_protocol_server: SSH-2.0-OpenSSH_8.9p1 Ubuntu-3ubuntu0.1
    encryption_algorithm: chacha20-poly1305@openssh.com
- stream: 15
  transport: tcp
  src: 172.16.225.48
  dst: 184.150.157.177
  src_port: 57380
  dst_port: 80
  ja4l_c: 45_128
  ja4l_s: 6252_58
  http:
  - ja4h: ge11nn030000_9ab90a797ba7_000000000000_000000000000
- stream: 22
  transport: tcp
  src: 172.16.225.48
  dst: 184.150.157.177
  src_port: 57396
  dst_port: 80
  ja4l_c: 50_128
  ja4l_s: 4272_58
  http:
  - ja4h: ge11nn030000_9ab90a797ba7_000000000000_000000000000
- stream: 33
  transport: udp
  src: 172.16.225.48
  dst: 142.251.32.74
  src_port: 51810
  dst_port: 443
  tls_server_name: signaler-pa.clients6.google.com
  ja4: q13d0312h3_55b375c5d22e_73e2d9e6cde6
  ja4s: q130300_1301_6bbbaf601ed8
- stream: 36
  transport: udp
  src: 172.16.225.48
  dst: 142.251.41.46
  src_port: 61861
  dst_port: 443
  tls_server_name: meet.google.com
  ja4: q13d0312h3_55b375c5d22e_73e2d9e6cde6
  ja4s: q130200_1301_234ea6891581
  ja4l_c: 169_128
  ja4l_s: 5389_57

